thunder/tests/test_tensor_subclass.py

```diff
@@ -202,17 +207,73 @@ def f(x: ScaleTensorSubclass, data: torch.Tensor, scale: torch.Tensor) -> ScaleT
     dtype = torch.float32
     shape = (2, 2)
     x = ScaleTensorSubclass(
-        make_tensor(shape, device=device, dtype=dtype),
+        make_tensor(shape, device=device, dtype=dtype, requires_grad=requires_grad),
         make_tensor((), device=device, dtype=dtype),
     )
-    data = make_tensor(shape, device=device, dtype=dtype)
+    y = ScaleTensorSubclass(
+        make_tensor(shape, device=device, dtype=dtype, requires_grad=requires_grad),
+        make_tensor((), device=device, dtype=dtype),
+    )
+
+    expected = f(x, y)
+    actual = jitted(x, y)
+    assert type(expected) is type(actual)
+    torch.testing.assert_close(expected, actual)
+    if requires_grad:
+        actual.mean().backward()
+
+    def g(x: ScaleTensorSubclass, data: torch.Tensor, scale: torch.Tensor) -> torch.Tensor:
+        y = EncapsulateXandScale.apply(data, scale)
+        out = x + y
+        return out
+
+    jitted = executor.make_callable(g)
+
+    x = ScaleTensorSubclass(
+        make_tensor(shape, device=device, dtype=dtype, requires_grad=requires_grad),
+        make_tensor((), device=device, dtype=dtype),
+    )
+    data = make_tensor(shape, device=device, dtype=dtype, requires_grad=requires_grad)
     scale = make_tensor((), device=device, dtype=dtype)
 
-    expected = f(x, data, scale)
+    expected = g(x, data, scale)
     actual = jitted(x, data, scale)
     assert type(expected) is type(actual)
-    torch.testing.assert_close((expected._x, expected._scale), (actual._x, actual._scale))
+    torch.testing.assert_close(expected, actual)
+    if requires_grad:
+        actual.mean().backward()
+
+
+@instantiate(
+    dtypes=(thunder.core.dtypes.float32,),
+    devicetypes=(thunder.core.devices.DeviceType.CUDA,),
+    executors=(TorchExecutor, TorchCompileCatExecutor, nvFuserExecutor, DynamoThunderExecutor),
+    decorators=(
+        pytest.mark.skipif(
+            not (TORCHAO_AVAILABLE and torch.cuda.get_device_capability() >= (8, 9)),
+            reason="Requires capability >= 8.9 and torchao",
+        ),
+    ),
+)
+def test_torchao_float8_linear(executor, device, _):
+    from torchao.float8 import convert_to_float8_training
+
+    batch_size, in_features, out_features = 16, 32, 64
+
+    device = torch.device("cuda")
+    dtype = torch.float32
+
+    model = nn.Linear(in_features, out_features, bias=False, device=device, dtype=dtype)
+    fp8_model = convert_to_float8_training(model)
+    x = make_tensor((batch_size, in_features), device=device, dtype=dtype)
+
+    expected = fp8_model(x)
+
+    jitted = executor.make_callable(fp8_model)
+    actual = jitted(x)
 
-    return_bsym: BoundSymbol = thunder.last_traces(jitted)[-1].bound_symbols[-1]
-    return_proxy = return_bsym.flat_args[0]
-    assert isinstance(return_proxy, SubclassTensorProxy)
+    if executor == DynamoThunderExecutor:
+        with pytest.raises(AssertionError):
+            torch.testing.assert_close(actual, expected)
```

Comment:
This failure doesn't feel easy to fix to me. So I made this into a script:

```python
import torch
import torch.nn as nn
from torchao.float8 import convert_to_float8_training
from thunder.dynamo import ThunderCompiler
from thunder.dynamo.splitter import SubgraphInfo
from thunder.tests.make_tensor import make_tensor


def main():
    batch_size, in_features, out_features = 16, 32, 64

    device = torch.device("cuda")
    dtype = torch.float32

    model = nn.Linear(in_features, out_features, bias=False, device=device, dtype=dtype)
    fp8_model = convert_to_float8_training(model)
    x = make_tensor((batch_size, in_features), device=device, dtype=dtype)
    expected = fp8_model(x)

    backend = ThunderCompiler()
    jitted = torch.compile(fp8_model, backend=backend)
    actual = jitted(x)

    backend.save_reproducer_to_folder("./debug_torchao_with_thunderfx", use_pytest_benchmark=True)
    print(f"{len(backend.subgraph_infos) = }")
    subgraph: SubgraphInfo
    for subgraph in backend.subgraph_infos:
        print(f"# {len(subgraph.thunder_compiled_fns) = }")

    torch.testing.assert_close(actual, expected)


if __name__ == "__main__":
    main()
```

note that https://github.com/pytorch/ao/pull/1339 is needed at the moment.

Below, I put the console output of the script above:
```
% python debug_thunderfx_torchao_fp8.py
/home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/thunder/dynamo/compiler.py:21: UserWarning: The ThunderCompiler is in active development and may not work as expected. Please report any issues you encounter to the Lightning Thunder team.
  warnings.warn(
len(backend.subgraph_infos) = 1
# len(subgraph.thunder_compiled_fns) = 0
Traceback (most recent call last):
  File "/home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/debug_thunderfx_torchao_fp8.py", line 34, in <module>
    main()
  File "/home/mkozuki/ghq/github.com/Lightning-AI/lightning-thunder/debug_thunderfx_torchao_fp8.py", line 30, in main
    torch.testing.assert_close(actual, expected)
  File "/home/mkozuki/ghq/github.com/crcrpar/pytorch/torch/testing/_comparison.py", line 1530, in assert_close
    raise error_metas[0].to_error(msg)
AssertionError: Tensor-likes are not close!

Mismatched elements: 388 / 1024 (37.9%)
Greatest absolute difference: 0.18639898300170898 at index (1, 61) (up to 1e-05 allowed)
Greatest relative difference: 1.9664803743362427 at index (10, 33) (up to 1.3e-06 allowed)
```

So it seems that `thunder.jit` isn't used for this program but the numeric is diverging.

Reply:
Can you check the result to see if they stay the same between different invocations. (Maybe due to low precision, the results could be different).

```python
expected = fp8_model(x)
actual = fp8_model(x)
torch.testing.assert_close(actual, expected)
```

Reply:
But please add a comment why expected and actual are both from calling the same model rather than one model and a reference.

Reply:
For thunderfx, I updated the test to check the parity between inductor and ThunderCompiler, not eager and ThunderFX.